### PR TITLE
split_at for Array

### PR DESCRIPTION
**Pull request by vbarrielle**
This PR introduces a `split_at`like family of functions for arrays. This is currently work in progress.

Status:
- [x] `axis_split_at`
- [x] `axis_split_at_mut`
- [x] tests

### CODE REVIEW

**Comment by bluss** [comment]
I think I prefer if we only provide axis split (not outer split, since it's a simple special case, isn't it?)

File: src/lib.rs [hunk 1]
```diff
@@ -1395,6 +1395,57 @@ impl<A, S, D> ArrayBase<S, D> where S: Data<Elem=A>, D: Dimension
...
+            let offset = self.dim.stride_offset_checked(&self.strides,
+                                                        &indices).unwrap();
+            unsafe {
+                self.ptr.offset(offset)
+            }
+        };
+
+        let mut dim_left = self.dim.clone();
+        dim_left.slice_mut()[axis] = index;
+        let left = ArrayView {
```
Comment (bluss) [inline src/lib.rs]: You can use `ArrayView::new_` here.

**Comment by vbarrielle** [comment]
Indeed it's just a special case, but I was wondering if we didn't want to have some symmetry with the `outer_iter`, `axis_iter` methods?

File: src/lib.rs (hunk 1 above)
Comment (vbarrielle) [inline src/lib.rs]: right, that's less boilerplate

**Comment by bluss** [comment]
The most flexible API is not the easiest to use. Again a view based `self -> (Self, Self)` function may be the easiest.

Consider a function like this:

`fn divide_and_conquer<'a, A>(array: ArrayView<'a, A, Ix>) { }`

The only way to pass on a view with the same lifetime `'a` is the `self` based methods I think. We want to be able to split an array view without tying the split result to the stack frame where the split happens.

This is the same way `.into_iter()` for an array view works, it is lifetime preserving unlike `.iter()`. So if you examine this problem, you find that maybe you want `into_inner_iter()` etc methods for array views too. (Lifetime-preserving iterator constructors). It's all quite complicated.

**Comment by bluss** [comment]
Consistency is good, but we need to stem the rising tide of a million methods too. Maybe I shouldn't have added `OuterIter` either.

**Comment by vbarrielle** [comment]
Yes I see your point with the lifetimes issue. I wanted to mimick the slice's API but I guess there is some special treatment to slices here.

One (probably ugly) way I found to circumvent that problem in other crates (sprs) was to have additional methods defined only on views which would return views based on the lifetime of the data, not the lifetime of self in the method call. I wish there were some sort of "lifetime specialization" to overcome this issue. Not sure I'm really clear here, my thoughts are a bit blurry and I think I lack the proper vocabulary to express them.

**Comment by bluss** [comment]
Yes! Slices are dsts, so they implement methods on the type `[T]`. If you look closely, you'll find lots of traces of this issue where it would be cool if an `ArrayView` could be a custom DST like that.

I think that `ArrayView::axis_split_at(self)` and `ArrayViewMut::axis_split_at(self)` is what we want here. We have to solve view methods not showing up in the docs, though.

**Comment by vbarrielle** [comment]
> We have to solve view methods not showing up in the docs, though.

If I remember correctly, I've had a similar issue in sprs, and the solution was to define the impl block on the full type, not the alias type. That was quite some time ago but it might be worth a try.

**Comment by bluss** [comment]
Yes, that will work

**Comment by bluss** [comment]
The for loop doesn't seem right.

Is this right: consider a 3 x 4 x 5 array (3 pages, 4 rows, 5 columns)

Split on axis = 2, index = 2 should produce two views of shapes
3 x 4 x 2 and 3 x 4 x 3

**Comment by vbarrielle** [comment]
I'm not sure I got your point here, that for loop is only supposed to help find the correct pointer offset, but the resulting shapes should be what you said.
However it's getting late here so it's quite probable I'm too sleepy to see the issue.

**Comment by bluss** [comment]
I'll review it more carefully. Some pen and paper doodling (that's when you know you're actually thinking..).

**Comment by vbarrielle** [comment]
I've switched to the view-by-value pattern and added some tests. I'll probably add a macro to factor out the code once again, and I need to test more complicated cases.

**Comment by bluss** [comment]
Awesome

File: src/lib.rs [hunk 2]
```diff
@@ -879,6 +879,51 @@ impl<'a, A, D> ArrayView<'a, A, D>
...
+            let mut indices = self.dim.clone();
+            for (ax, ind) in indices.slice_mut().iter_mut().enumerate() {
+                if ax != axis {
+                    *ind = 0;
+                }
+                else {
+                    *ind = index;
+                }
+            }
+            let offset = self.dim.stride_offset_checked(&self.strides,
```
Comment (bluss) [inline src/lib.rs]: There's the `stride_offset` method which does this without bounds checking.

File: src/lib.rs [hunk 3]
```diff
@@ -879,6 +879,51 @@ impl<'a, A, D> ArrayView<'a, A, D>
...
+        let right_ptr = if index == self.shape()[axis] {
+            self.ptr
+        }
+        else {
+            let mut indices = self.dim.clone();
+            for (ax, ind) in indices.slice_mut().iter_mut().enumerate() {
+                if ax != axis {
+                    *ind = 0;
+                }
+                else {
```
Comment (bluss) [inline src/lib.rs]: Can you put the `else` on the same line as the `}`? Thanks.

File: src/lib.rs (hunk 2 above)
Comment (bluss) [inline src/lib.rs]: On a closer look, we can just skip the whole for loop, can't we? Using stride * index directly (function `stride_offset` in dimension for example).

**Comment by vbarrielle** [comment]
I've fixed the formatting issue, replaced the for loop by an appropriate call to `dimension::stride_offset`, and added more tests.

I'm no longer sure I should try and factor both impls in a macro, they're quite short now. As you wish :)

**Comment by bluss** [comment]
Awesome, this is good. I'll merge the Axis PR too.

**Comment by bluss** [comment]
Before I release this, what do you think of renaming it?

`pub fn split_at(self, axis: Axis, index: Ix)`

versus

`pub fn split_at(self, axis: Axis, index: Ix)`

I bring this up since `split_at` seems nice an unambiguous too.

**Comment by vbarrielle** [comment]
Yes `split_at` seems nice and unambiguous, I don't think we could ever have a split_at taht does not require an axis argument.

**Comment by bluss** [comment]
Done.

**Pull request merged**